### ISSUE TITLE
[Flang] Incorrect diagnostic on defined assignment when first argument is unlimited polymorphic

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#124621]
Consider the following code:
```
module m

   type base
      integer(4) :: i = -999
      character(3) :: c = 'xxx'
      contains
         procedure, pass(b) :: U_base
         generic :: assignment(=) => U_base
   end type

   contains

   subroutine U_base ( a, b )
      class(*), intent(out) :: a
      class(base), intent(in)   :: b

   end subroutine

end module
```

Flang currently issues an error as:
```
./t.f:8:38: error: Defined assignment subroutine 'u_base' conflicts with intrinsic assignment
           generic :: assignment(=) => U_base
                                       ^^^^^^
./t.f:13:15: Declaration of 'u_base'
     subroutine U_base ( a, b )
                ^^^^^^

```

The code seems conforming to me, and all ifort, gfortran and XLF compiled it successfully.

**Comment by klausler** [comment]
F'2023 15.4.3.4.3 p2: "Either the second argument shall be an array whose rank differs from that of the first argument, the declared types and kind type parameters of the arguments shall not conform as specified in Table 10.8, or the first argument shall be of derived type."

The unlimited polymorphic first argument should render the argument as not conforming "as specified in Table 10.8", so this usage is valid.

Will fix.
